docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but many of these features and changes are motivated by external use cases, which need enablement. Not all of these use cases will be completely enabled by this proposal.
+
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a remote container/vm monitoring tool (e.g. bosh monitor), I want to be able to determine the health of the component container/vm I am monitoring.
+
+Note: Self-healing requires knowing how to deploy the cluster components, which is different for each cloud platform and changes based on component configuration.
+
```

Comment:
As far as client, i meant "programmatic clients that expect to work generically with named resource types" (so kubectl, operating across multiple API groups, or a generic Kube console that supported other Kube resource APIs that aren't in the primary group) as well as "other APIs that would be consumed by clients, like logging and metrics" as well as "other web endpoints that are generally useful for administration of the cluster, like logging consoles and metrics consoles".

The reason why ingress is mentioned is that these components will live on the cluster, but to be visible to clients, must be exposed outside the cluster (be nodeport, load balanced service, or ingress).   To a client, the method of ingress/egress, just the outcome (there is an address the client can talk to).

Calling "get services in a namespace" isn't enough (as per the discussions in this thread earlier), because exposing a service to all users via an API call may separate concerns.

So the use cases you described are more "here's how these can be used inside the cluster" or "how services are exposed", but I'm specifically referring to the desire of clients to get a transformed list of "published" APIs.  Most admins would consider those "components" (my logging component, my build service that exposes the build resource component) and this proposal needs to at least talk to how they are relevant (I think the mechanisms you describe are sufficient to register and monitor them, but it's the step beyond that needs at least a minor pointer).